source/extensions/filters/network/mongo_proxy/bson_impl.cc

```diff
@@ -101,7 +101,7 @@ std::string BufferHelper::removeString(Buffer::Instance& data) {
   }
 
   char* start = reinterpret_cast<char*>(data.linearize(length));
-  std::string ret(start);
+  std::string ret(start, length);
```

Comment:
awesome, thanks. Can I also ask for a unit test that fails if you don't have this change?

Reply:
Due to this change, the following unit test fails 
```
MongoCodecImplTest.Query
MongoCodecImplTest.Command
MongoCodecImplTest.CommandReply
```
before that only coverage build fails on CI.

Reply:
I think this change looks correct, but I must not understand it well enough, and there's some subtlety I'm missing. Like maybe there are valid embedded null characters expected in this block and we need to use strlen or similar to terminate them?

This code-block dates back from the initial import of Envoy into open-source from Lyft (#2). Maybe @mattklein123 can point us to someone who knows this code better?

Reply:
See http://bsonspec.org/spec.html. This code as written is not robust to bad data. This also looks correct to me. I can't say why the tests are failing but something needs to be fixed and new tests added.

Reply:
May be below log will point to actual cause for the failure.
```
unknown file: Failure

Unexpected mock function call - returning directly.
    Function call: decodeCommand_(@0x7ffc38809ac0 8-byte object <C0-6A D9-BF 1E-4D 00-00>)
Google Mock tried the following 1 expectation, but it didn't match:

test/extensions/filters/network/mongo_proxy/codec_impl_test.cc:396: EXPECT_CALL(callbacks_, decodeCommand_(Pointee(Eq(command))))...
  Expected arg #0: points to a value that is equal to 144-byte object <A8-A0 E1-37 77-55 00-00 0F-00 00-00 19-00 00-00 68-A1 E1-37 77-55 00-00 D0-1E C
2-BF 1E-4D 00-00 0D-00 00-00 00-00 00-00 54-65 73-74 20-64 61-74 61-62 61-73 65-00 34-22 40-AB DD-BF 1E-4D 00-00 ... 00-00 00-00 70-22 3A-20 E0-1C A1-
BF 1E-4D 00-00 40-38 DD-BF 1E-4D 00-00 80-AD A1-BF 1E-4D 00-00 40-23 D9-BF 1E-4D 00-00 20-A7 DD-BF 1E-4D 00-00 20-A7 DD-BF 1E-4D 00-00 01-00 00-00 00-
00 00-00>
           Actual: 8-byte object <C0-6A D9-BF 1E-4D 00-00>, which points to 8-byte object <68-A1 E1-37 77-55 00-00> (of type Envoy::Extensions::Networ
kFilters::MongoProxy::CommandMessage)
         Expected: to be called once
           Actual: never called - unsatisfied and active
```

Reply:
The issue (I think) is that the length in the BSON string might be greater than the data that is linearized. You need to either treat this as an error case or truncate it appropriately. I would treat it as an error case.

Reply:
Truncated '\0'.

Reply:
I've lost track of why this change is being made. What you have now is I think identical to the original code, achieving the same result in one more step.

It would be helpful if you included a unit test that fails in the original state but passes with your source changes.